lib/runner/cli/cli.js

```diff
@@ -236,7 +236,6 @@ class CliRunner {
     const {start_process} = webdriver;
     const shouldStartSeleniumProcess = WebDriver.isUsingSeleniumServer(this.test_settings);
     const isTestWorker = this.testRunner.isTestWorker();
-    const testWorkersEnabled = this.isTestWorkersEnabled();
```

Comment:
do we not have test case for this ?

Reply:
Yeah, this is likely to create other problems.

Reply:
>do we not have test case for this that we need to remove?

nope

Reply:
>Yeah, this is likely to create other problems.

can you please briefly explain what sort of problems can occur? Are these edge cases we are talking about here.

IMO, we should not care about `test_workers` and `single` test files. We should let Webdriver spawn the driver, be it we have multiple files, or single file or `test_worker` is enabled

Reply:
There is a big difference between a test worker running in its own process and a single test file. If you run a single test file then you always need to start the driver (assuming that start_process is true), but if you run in test workers mode and use the selenium server, then you only need to start the driver in the main process. If you run in test workers mode and you only use geckodriver, then you need to start the driver for each test worker, using a random port.

And there are also considerations when running in both parallel mode with multiple environments and test workers, which I don't remember directly. 

So I'm not sure why this change is necessary here. Is this what's causing the reported issue?

Reply:
Thanks for the explanation.

> Is this what's causing the reported issue?

yes 😅

Reply:
I ran few benchmarks with and without my fix and below are the result:

#### Without my fix - 100 iterations - (2 test files)
452.11s user 174.10s system 77% cpu 13:24.55 total
480.48s user 187.08s system 85% cpu 13:04.55 total
474.78s user 181.79s system 84% cpu 12:56.56 total
959.61s user 363.67s system 106% cpu 20:43.30 total
1028.01s user 385.57s system 113% cpu 20:42.68 total

**Avg: 678.6**

#### With my fix - 100 iterations - (2 test files)
473.14s user 182.98s system 83% cpu 13:09.43 total
497.31s user 189.31s system 79% cpu 14:21.15 total
509.67s user 195.94s system 84% cpu 13:57.91 total
461.65s user 177.60s system 74% cpu 14:23.22 total
451.70s user 171.64s system 81% cpu 12:48.73 total

**Avg: 478.2**

#### Without my fix - 50 iterations - (7 test files)
991.03s user 382.04s system 182% cpu 12:32.99 total

### With my fix - 50 iterations - (7 test files)
1023.16s user 394.75s system 176% cpu 13:24.89 total

@beatfactor I think we can go ahead with this modification of the `singleSourceFile` function?

> if you run in test workers mode and use the selenium server, then you only need to start the driver in the main process. 

In https://github.com/nightwatchjs/nightwatch/blob/da83b77d28d07583ee6a65dac3dbbeb67fa4908a/lib/runner/cli/cli.js#L242, we blindly return true in case of selenium, so we aren't checking for `test_workers`, and `testWorkersMode` will be true in any case, if multiple files are present

Reply:
I'm not in favour of the changes in the "singleSourceFile" for the reasons I've already explained. Even if the impact is negligible in our benchmarks, it is conceptually the wrong way to do it and we should avoid it. 

I think what we can do is read the test source before the startWebDriver() call, although this sounds a bit too risky for how trivial this bug actually is. I'd say to make the change in v2 and leave this as it is because it's a very low impact bug which requires risky changes or ugly workarounds.

Also, since you removed the `testWorkersEnabled` check, the `singleSourceFile` check is not performed anymore.